WIP Parsers tests

streeg opened the pull request.

lfcnassif commented:
Hi @streeg, firt basic review:
- do not use import *
- As you noticed, BasicProps class static fields should not be changed. I will refactor them on #18;
- SevenZipParser: I think you could check the md5 of extracted files (like MboxParser below) to check if the files are being fully extracted. It is also important to test the inner path of extracted files;
- MboxParser: you are parsing manually the mbox file instead of using MBoxParser. You should do something like this:
```
context.set(Parser.class, new AbstractParser(){
   //this will be called for embedded emails
    public void parse(InputStream stream, ContentHandler handler, Metadata metadata, ParseContext context) throws IOException, SAXException, TikaException {
        // compute md5 hash of stream and store in metadata or other field of this class,
        // so you can check if mbox parser is extracting mails correctly.
        // you can also count the number of mails extracted
    }
});
```
I think the testing of inner mails extracted by MboxParser could be done just in RFC822ParserTests
- XMLParser: I think metadata extracted values should be tested instead of checking them printed in bodyContentHandler

lfcnassif commented:
No need to test MSAccess now, possibly will be removed: https://github.com/sepinf-inc/IPED/issues/461

streeg commented:
I'll change the import *.

I imagined that. Had changed for testing, was not my intention to push.

I'll take a look and test the md5 of files in SevenZipParserTest and in RARParserTest.

I'll also refactor the mbox parser and check xml metadata instead, thanks for the tip.

streeg commented:
I believe that the XMLParser, even with the master changes, still not filling the metadata content properly.

lfcnassif commented:
It is possible, please take a closer look and try to fix it if not working in another branch/PR.

lfcnassif commented:
About commit da4b867, MD5 is still not being computed for embedded items, you are trusting in provided CRC (different from MD5, do not store it as MD5) value from RAR header. You should read all subitem inputStream content and compute MD5 in the test and compare it to the known MD5 of the test file (computed externally using other tool).

I also think your changes to MBoxParser in da4b867 are not needed, you changed the parser to adapt it to the test. Parsers could be fixed and improved (in other tickets), but I think those changes are not needed.

lfcnassif commented:
MBoxParser tests could only count the number of emails extracted and if the MD5 of extracted emails matches the expected values. Email header metadata (from, to...) should be tested in RFC822Parser tests

streeg commented:
I'll make the necessary changes to MboxParserTests and then compute and compare the md5 for input stream sub items.

streeg commented:
Ok, some weird thing happened to ImageOCRMetadataParser. The metadata for the tested BMP file returns "**BitsPerSample=8 8 8 8**". The test pass for java 11 but fails with java 8 and java 14 expecting "**BitsPerSample=8 8 8**", is this related with the parser? (My local java version is 1.8, same used in the eclipse IDE where the tests passed)
![testbmp](https://user-images.githubusercontent.com/22523306/115917163-30e95580-a44c-11eb-8934-65efd70b6624.png)
 
I have tried to use digest() with the input stream for the sub item for capturing the md5 but had unsuccessful results any tips?
![digest](https://user-images.githubusercontent.com/22523306/115917472-ad7c3400-a44c-11eb-8fc5-9dc8bc910eeb.png)

lfcnassif commented:
The BMP test failure on java 8/14 is really strange, no initial ideia about its cause. About MD5, you cannot do `is.toString()` to read inputStream bytes, please read carefully all java InputStream API, it is used through all the project:
https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html

And I suggest using DigestUtils instead of MessageDigest, that is much simpler to use.

streeg commented:
I have succeeded getting the md5 from files using the MessageDigest. I'm having trouble now with the handler. Whenever I declare the string to take the InputStream digest, the handler simply does not parse the body content. 
![digest](https://user-images.githubusercontent.com/22523306/116137195-9dad5b80-a6a9-11eb-9af7-04fb37cbdbae.png)
I tried to debug without success... I'll upload with the handler tests commented.

lfcnassif commented:
> I have succeeded getting the md5 from files using the MessageDigest. I'm having trouble now with the handler. Whenever I declare the string to take the InputStream digest, the handler simply does not parse the body content.
> ![digest](https://user-images.githubusercontent.com/22523306/116137195-9dad5b80-a6a9-11eb-9af7-04fb37cbdbae.png)
> I tried to debug without success... I'll upload with the handler tests commented.

When you compute the MD5, you are consuming all the inputStream bytes, so no data is left to be parsed by the parser. You must use inputStream mark() reset() methods, they are expained in the API docs I linked before. Check the markSupported() return value, if it is false, you will need to wrap the inputStream in a BufferedInputStream before marking and before MD5 calculation.

streeg commented:
Solved by using my abstrackpackage instead of hardcoding the parser. Sorry, my mistake. Thanks for the tip!!

lfcnassif commented:
> Solved by using my abstrackpackage instead of hardcoding the parser. Sorry, my mistake. Thanks for the tip!!

Good. I also confused things, you are computing MD5 for subitems, not for the parent item, sorry.

streeg commented:
Ok, since I last posted here, I have finished all the things that have been proposed except the XMLParser. I added the OutlookPSTParser as well started the RegistryParser test. I'll remove the SOFTWARE later due to its size of 80mb! Sorry, didn't notice. Also, what should I do to build here in git when testing, due to the parser use of external application (like RegRipper)? I have only succeeded when testing locally.

lfcnassif reviewed (changes requested):
Please revert the flags changed on OutlookPSTParser. Use the setter methods to set them from the unit test.

lfcnassif reviewed (changes requested):
I think you will need to update the CI configuration file (https://github.com/sepinf-inc/IPED/blob/master/.github/workflows/maven.yml) to install regripper into the CI ubuntu environment. Luckly it will be automatically recognized by the parser.

lfcnassif commented:
> I believe that the XMLParser, even with the master changes, still not filling the metadata content properly.

I will take a look at this.

lfcnassif commented:
> Ok, since I last posted here, I have finished all the things that have been proposed except the XMLParser. I added the OutlookPSTParser as well started the RegistryParser test. I'll remove the SOFTWARE later due to its size of 80mb! Sorry, didn't notice. Also, what should I do to build here in git when testing, due to the parser use of external application (like RegRipper)? I have only succeeded when testing locally.

It would be better to use a smaller SYSTEM test file too.

lfcnassif commented:
> > I believe that the XMLParser, even with the master changes, still not filling the metadata content properly.
> 
> I will take a look at this.

Done. IPED's XMLParser is using Tika's HTMLParser internally instead of Tika's XMLParser (this works to extract dublin core metadata). I "think" HTMLParser is less strict and works better with mal formed files. Actually Tika app (or server, don't remeber) also uses HTMLParser for XML files. I will need to review and test the differences between the two to decide if we can switch without regressions.

streeg commented:
I'm currently doing the PackageParser tests, for now, whenever I try to open a jar file using the parser it works fine. I can read all the content, extract metadata and its content from the handler. I just noticed something weird. When the parser tries to set its content_type, instead of setting as JAR, it is setting as ZIP, and since you can manipulate (extract the files) jar as zip, I can still do all the things I want to as it was a ZIP file. 

<img width="400" alt="2021-04-30 16_06_38-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_sp_gpinf_indexer_parsers" src="https://user-images.githubusercontent.com/22523306/116743170-eda06100-a9ce-11eb-9bd2-a7de654fa9fd.png">
<img width="800" alt="2021-04-30 16_07_49-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_sp_gpinf_indexer_parsers2" src="https://user-images.githubusercontent.com/22523306/116743180-ef6a2480-a9ce-11eb-97d7-f3b7507b8a5a.png">

I think that is not intended... Or is it? 

*I'll also add more test methods in all PackageParsers (sevenzip and rar as well), to test not only correct things but all the remains methods of the Parsers that are currently not being tested.

lfcnassif commented:
> I'm currently doing the PackageParser tests, for now, whenever I try to open a jar file using the parser it works fine. I can read all the content, extract metadata and its content from the handler. I just noticed something weird. When the parser tries to set its content_type, instead of setting as JAR, it is setting as ZIP, and since you can manipulate (extract the files) jar as zip, I can still do all the things I want to as it was a ZIP file.
> 
> <img alt="2021-04-30 16_06_38-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_sp_gpinf_indexer_parsers" width="400" src="https://user-images.githubusercontent.com/22523306/116743170-eda06100-a9ce-11eb-9bd2-a7de654fa9fd.png">
> 
> <img alt="2021-04-30 16_07_49-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_sp_gpinf_indexer_parsers2" width="800" src="https://user-images.githubusercontent.com/22523306/116743180-ef6a2480-a9ce-11eb-97d7-f3b7507b8a5a.png">
> 
> I think that is not intended... Or is it?
> 
> *I'll also add more test methods in all PackageParsers (sevenzip and rar as well), to test not only correct things but all the remains methods of the Parsers that are currently not being tested.

I think this isn't totally wrong, since a jar is a zip. Not sure if the parser can detect the custom type parsing just the inputstream, maybe checking some entry names for a MANIFEST for other jar specific entries. But actually, this code is useless for Iped, because the contentType is already set by a detector run before, that uses name and extension hints (and I think it looks for entry names too) to differentiate between a general zip, jar, or other zip based files...

lfcnassif commented:
Maybe we can remove this code that set the contentType from the parser, instead of ducplicating Tika's ZIP detector logic.

lfcnassif reviewed (commented):
Seems to me the encodings of the samples could be wrong. For a one encoding file, just open notepad, write the text with diacritics and special chars, go to "save as" and choose the appropriate encoding.

To create files with mixed encodings, you will have to concatenate files like above in binary form, it will be difficult to use any text editor to save text using multiple encodings.

lfcnassif commented:
@streeg #527 was merged, please revert your changes in BasicProps class and merge master into this branch, that should fix your issues.

streeg commented:
> @streeg #527 was merged, please revert your changes in BasicProps class and merge master into this branch, that should fix your issues.

I'll do that!

streeg commented:
> @streeg #527 was merged, please revert your changes in BasicProps class and merge master into this branch, that should fix your issues.

So, after the merge locally, i'm getting:
```
java.lang.ExceptionInInitializerError
Caused by: java.util.MissingResourceException: Can't find bundle for base name iped-parsers-messages, **locale pt_BR.**
```
Have you tested for this locale?

streeg commented:
> I didn't test the changes in this branch. When are you getting this error, building or running the app? What is the full stacktrace?

Alguns testes estão passando normalmente. Outros testes quebram com esse log que eu mandei, geralmente quando o parser chama o getString() do message. 

Aqui outro log que só quebra quando chama o getString no parseSubItem do SevenZipParser: 

```
java.lang.ExceptionInInitializerError
	at dpf.sp.gpinf.indexer.parsers.util.Util.<clinit>(Util.java:35)
	at dpf.sp.gpinf.indexer.parsers.SevenZipParser$MyExtractCallback.parseSubitem(SevenZipParser.java:271)
	at dpf.sp.gpinf.indexer.parsers.SevenZipParser$MyExtractCallback.setOperationResult(SevenZipParser.java:246)
	at dpf.sp.gpinf.indexer.parsers.SevenZipParser.parse(SevenZipParser.java:142)
	at dpf.sp.gpinf.indexer.parsers.SevenZipParserTest.testSevenZipRAR5Parsing(SevenZipParserTest.java:26)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.util.MissingResourceException: Can't find bundle for base name iped-parsers-messages, locale pt_BR
	at java.util.ResourceBundle.throwMissingResourceException(Unknown Source)
	at java.util.ResourceBundle.getBundleImpl(Unknown Source)
	at java.util.ResourceBundle.getBundle(Unknown Source)
	at dpf.sp.gpinf.indexer.parsers.util.Messages.getString(Messages.java:20)
	at dpf.sp.gpinf.indexer.parsers.IndexerDefaultParser.<clinit>(IndexerDefaultParser.java:76)
	... 24 more
```

![2021-05-10 15_22_38-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_sp_gpinf_indexer_parsers](https://user-images.githubusercontent.com/22523306/117706413-a3279d00-b1a3-11eb-959c-c763b65e3f3a.png)

lfcnassif commented:
Ok this is because all bundles (including english) were moved to an external folder. This should break debugging too. Will add a temp workaround to return the key string if the bundle is not found and will research later if it is possible to keep the bundles in their original locations and just externalize them while building the application release.

streeg commented:
I was testing VCardParser and i'm getting an error when it tries to parse a VCard with profile picture as seen below: 
![errowhenvcardhasthumb](https://user-images.githubusercontent.com/22523306/118538807-0aee6280-b725-11eb-90e3-f34c5ff57be6.png)

I'm getting the following log:
```

java.lang.NullPointerException
	at java.util.Base64$Encoder.encode(Unknown Source)
	at java.util.Base64$Encoder.encodeToString(Unknown Source)
	at dpf.mg.udi.gpinf.vcardparser.VCardParser.extractMetadata(VCardParser.java:157)
	at dpf.mg.udi.gpinf.vcardparser.VCardParser.parse(VCardParser.java:93)
	at dpf.mg.udi.gpinf.vcardparser.VCardParserTest.testVCardCompleteInfoParser(VCardParserTest.java:81)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)


```

lfcnassif commented:
> I was testing VCardParser and i'm getting an error when it tries to parse a VCard with profile picture as seen below:
> ![errowhenvcardhasthumb](https://user-images.githubusercontent.com/22523306/118538807-0aee6280-b725-11eb-90e3-f34c5ff57be6.png)
> 
> I'm getting the following log:
> 
> ```
> 
> java.lang.NullPointerException
> 	at java.util.Base64$Encoder.encode(Unknown Source)
> 	at java.util.Base64$Encoder.encodeToString(Unknown Source)
> 	at dpf.mg.udi.gpinf.vcardparser.VCardParser.extractMetadata(VCardParser.java:157)
> 	at dpf.mg.udi.gpinf.vcardparser.VCardParser.parse(VCardParser.java:93)
> 	at dpf.mg.udi.gpinf.vcardparser.VCardParserTest.testVCardCompleteInfoParser(VCardParserTest.java:81)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
> 	at java.lang.reflect.Method.invoke(Unknown Source)
> 	at junit.framework.TestCase.runTest(TestCase.java:177)
> 	at junit.framework.TestCase.runBare(TestCase.java:142)
> 	at junit.framework.TestResult$1.protect(TestResult.java:122)
> 	at junit.framework.TestResult.runProtected(TestResult.java:142)
> 	at junit.framework.TestResult.run(TestResult.java:125)
> 	at junit.framework.TestCase.run(TestCase.java:130)
> 	at junit.framework.TestSuite.runTest(TestSuite.java:241)
> 	at junit.framework.TestSuite.run(TestSuite.java:236)
> 	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
> 	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
> 	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
> ```

Seems getData() is returning null. Please open an issue and attach the sample file so we can reproduce and try to fix it.

lfcnassif commented:
@streeg I reverted some logic changes in some parsers after a brief review and removed some commented code left behind. Please I ask you to review all changed parsers for logic behavior changes. If we need to change some logic, first an issue must be opened to document and track it. No problem about new setters and getters created.

About MboxParser tests, they are failing after I reverted things. I ask you to adjust the test to just count the number of subitems/emails extracted and compare their MD5 to known values. Values of TO, FROM, DATE and other email headers should be tested in RFC822ParsersTests.

lfcnassif commented:
Also translate all your comments to english, this needs to be done for all code base, let's not increase this future work.

streeg commented:
I'll do that!

lfcnassif commented:
> About MboxParser tests, they are failing after I reverted things. I ask you to adjust the test to just count the number of subitems/emails extracted and compare their MD5 to known values. Values of TO, FROM, DATE and other email headers should be tested in RFC822ParsersTests.

MBoxParser is like any other container parser (PackageParser, SevenZipParser...), it extracts subitems and send them to other parsers. So we need to check the number of subitems extracted, if they were extracted correctly (e.g. checking MD5), if internal path of subitems (internal tree structure) are decoded fine (it does not exist in mbox) and check some subitem metadata **kept by parent container**. There is no such metadata kept by mbox files, all metadata are into the eml subitem files and will be extracted by RFC822Parser. But other containers keep some metadata about subitems e.g. zip, rar, 7z files: MAC dates of subitems, CRC, tree structure, etc

streeg commented:
Accidentally clicked to request review... Sorry, I was trying to click to see the recently changes!

lfcnassif reviewed (changes requested):
review all logic changes outside tests and fix MBoxParser tests as asked before, thanks.

streeg commented:
I just checked the changes made to the parsers. It hasn't changed how they work. The changes I made previously (which were removed), were in large majority just to be able to test one method or the other. As for MBoxParser, I changed it to capture and check only the md5, the type of content and I took advantage of the fact that the parser itself already adds the subject and I checked it too.

lfcnassif commented:
> I just checked the changes made to the parsers. It hasn't changed how they work. The changes I made previously (which were removed), were in large majority just to be able to test one method or another. As for MBoxParser, I changed it to capture and check only the md5, the type of content and I took advantage of the fact that the parser itself already adds the subject and I checked it too.

Thanks for the review! Actually RFC822Parser logic was changed when isUtf8 boolean was made a class attribute. The same parser instance is reused to parse several files, so that field could be set when parsing one file and used for other.

Long history about MBoxParser: Subject (converted to Name later) and Created metadata set by MboxParser actually are a workaround because "leaf" parsers (like RFC822Parser) cannot change basic properties currently, just container parsers can set them for subitems. MboxParser ideally should not call RFC822Parser (parsing the whole email) just to populate those basic properties, because RFC822Parser is called again later, running twice. Those metadata (subject and created) could be tested for now, because they are set. But adding unneeded metadata (to, from, date) to the workaround is not necessary, so that's why I removed them. Anyway, that changed behavior/results and we need to open an issue to list the changes in the release notes if we need to change/fix some parser.

streeg commented:
Hey @lfcnassif I'm trying to create the tests for winXTimelineParser. Whenever I try to parse using the parser it returns 
```
org.apache.tika.exception.TikaException: SQLite Win10Timeline parsing exception
	at dpf.inc.sepinf.winx.parsers.WinXTimelineParser.parse(WinXTimelineParser.java:132)
	at dpf.inc.sepinf.winx.parsers.WinXTimelineParserTest.testWinXTimelineParser(WinXTimelineParserTest.java:31)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: org.sqlite.SQLiteException: [SQLITE_ERROR] SQL error or missing database (no such table: Activity_PackageId)
	at org.sqlite.core.DB.newSQLException(DB.java:1012)
	at org.sqlite.core.DB.newSQLException(DB.java:1024)
	at org.sqlite.core.DB.throwex(DB.java:989)
	at org.sqlite.core.NativeDB.prepare_utf8(Native Method)
	at org.sqlite.core.NativeDB.prepare(NativeDB.java:134)
	at org.sqlite.core.DB.prepare(DB.java:257)
	at org.sqlite.jdbc3.JDBC3Statement.executeQuery(JDBC3Statement.java:66)
	at dpf.inc.sepinf.winx.parsers.WinXTimelineParser$1.<init>(WinXTimelineParser.java:212)
	at dpf.inc.sepinf.winx.parsers.WinXTimelineParser.runQuery(WinXTimelineParser.java:209)
	at dpf.inc.sepinf.winx.parsers.WinXTimelineParser.parse(WinXTimelineParser.java:108)
```

I believe the table names are not being populated. I tried to debug today and so far didn't understand what is causing this. Not only that, but I know for sure that the file being tested contains the tables. Should I open an issue and report this?

lfcnassif commented:
Not sure what's going on, have you tried to process the file using the full application? iped.exe -d folder_with_db -o output

streeg commented:
> iped.exe -d folder_with_db -o output

I have tried to open the .db using sqlite3 in my powerShell. I'll try to use iped.

lfcnassif commented:
@streeg I "think" I know what's going on, will test here to confirm the solution.

streeg commented:
> @streeg I "think" I know what's going on, will test here to confirm the solution.

Just saw #557. I think that will indeed fix the problem. Thanks!

streeg commented:
Hi, it's me yet again... Another day, another problem. I'm building the UsnJrnlParserTests now. When I try to parse using iped, I'm able to get all the information, but when I try to test using JUnit I'm getting NullPointerException:  

```
java.lang.NullPointerException
	at dpf.inc.sepinf.UsnJrnl.UsnJrnlParser.parse(UsnJrnlParser.java:227)
	at dpf.inc.sepinf.UsnJrnl.UsnJrnlParserTest.testUsnJrnlParsing(UsnJrnlParserTest.java:32)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

This led me to believe that the problem was related with the way the parser were receiving its content, like yesterday with WinXTimelineParser. However, when debugging I noticed that when the parser tries to create a _IItemSearcher searcher_ and a _IItemBase item_, it is receiving NULL from context.get(_IItemSearcher.class_) and  from context.get(_IItemBase.class_). I know the context is not null as when I get the _SupportedTypes_ it returns as expected. 

Here is a screenshot of the debugger. 
![2021-05-21 17_01_27-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_inc_sepinf_UsnJrnl_UsnJr](https://user-images.githubusercontent.com/22523306/119192262-37bcb580-ba56-11eb-842a-f5575afd30d5.png)

This issue of IItem Classes not receiving context (receiving NULL) has happened before when I tried to create EdgeWebCacheParser. I didn't say anything about because the test still in development as external software is needed. 

Any thoughts in what could it be?

lfcnassif commented:
It is similar to WinXTimelineParser issue. UsnJrnlParser assumes it is being called in IPED pipeline, where the context is filled with some objects implementing IItemBase and IItemSearcher, you will need to populate the context with some objects implementing those interfaces. IItemSearcher can return just an empty result list. IItemBase should implement the used getStream() method, others can be left no op,

lfcnassif commented:
The parser could throw a better exception instead of NPE, but it needs to retrieve those objects from context to work properly and to test its functionality.

streeg commented:
> It is similar to WinXTimelineParser issue. UsnJrnlParser assumes it is being called in IPED pipeline, where the context is filled with some objects implementing IItemBase and IItemSearcher, you will need to populate the context with some objects implementing those interfaces. IItemSearcher can return just an empty result list. IItemBase should implement the used getStream() method, others can be left no op because they are not used by the parser.

Ok, I have created two objects for _IItemSearcher_ and _IItemBase_. I have inserted these objects in the _context_ with _set_ method. So far,  _searcher_ and _item_ are no longer **_null_**. 

![2021-05-24 15_55_38-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_inc_sepinf_UsnJrnl_UsnJr](https://user-images.githubusercontent.com/22523306/119394441-8c576f00-bca8-11eb-9204-e3d15418d25e.png)

I understand that I am the one who should be filling the item's information, but how? I mean, how should I use the _getStream_ method to get an instance of the _Stream_ as a _SeekableInputStream_?

![2021-05-24 15_58_01-eclipse-workspace - iped-parsers-impl_src_test_java_dpf_inc_sepinf_UsnJrnl_UsnJr](https://user-images.githubusercontent.com/22523306/119394680-dccecc80-bca8-11eb-8a71-ab576765cc0e.png)

lfcnassif commented:
Try something like this in your implementation:

```
public SeekableInputStream getStream() throws IOException {
    return new SeekableFileInputStream(test_file);
}
```
And it is better if your IItemSearcher implementation returns an empty list in search() method instead of null.

streeg commented:
Hey @lfcnassif do you know where I can get samples for UFEDParser? If I'm not mistaken, UFED is paid software on which I don't have a license to be able to generate my own test files. If you don't have the files, do you know of anyone on staff who can help me? Usually I have been ordering the files via email from the parser authors.

lfcnassif commented:
Hum... actually the UFEDChatParser builds html reports of chats based on info extracted from UFED UFDR files by UfedXmlReader class. So actually don't need the UFED UFDR file to test it, just to generate input to the parser like is done by UfedXmlReader class. Maybe there are other easier tests to implement before this...

streeg commented:
> Hum... actually the UFEDChatParser builds html reports of chats based on info extracted from UFED UFDR files by UfedXmlReader class. So actually you don't need the UFED UFDR file to test it, just to generate input to the parser like is done by UfedXmlReader class. Maybe there are other easier tests to implement before this...

All right, I'll work on others. I'm postponing some like whatsapp and telegram too.

streeg commented:
Hi @lfcnassif, I am starting Telegram Parser Tests. 

I am receiving the following error whenever I try to parse my cache4.db file:

```
java.lang.ClassNotFoundException: telegramdecoder.DecoderTelegram
	at java.net.URLClassLoader.findClass(Unknown Source)
	at java.lang.ClassLoader.loadClass(Unknown Source)
	at sun.misc.Launcher$AppClassLoader.loadClass(Unknown Source)
	at java.lang.ClassLoader.loadClass(Unknown Source)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Unknown Source)
	at dpf.ap.gpinf.telegramextractor.TelegramParser.parseTelegramDBAndroid(TelegramParser.java:131)
	at dpf.ap.gpinf.telegramextractor.TelegramParser.parse(TelegramParser.java:454)
	at dpf.ap.gpinf.telegramextractor.TelegramParserTest.testTelegramParser(TelegramParserTest.java:67)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)

```
I know this missing class is in the dependencies. When I click to solve, it asks me to point to the class. When I do, it keeps giving this error.
![2021-06-09 17_09_08-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_ap_gpinf_telegramextract](https://user-images.githubusercontent.com/22523306/121422268-705cfa00-c945-11eb-801a-03b52410befe.png)

![2021-06-09 17_10_06-eclipse-workspace - iped-parsers-impl_src_main_java_dpf_ap_gpinf_telegramextract](https://user-images.githubusercontent.com/22523306/121422376-908cb900-c945-11eb-8371-b5fa11ef465b.png)

What should I do?

Also, the cache4.db file is the one I should be parsing, right? (Android version) 

ps: I'll remove unnecessary files later.

lfcnassif commented:
> What should I do?

I will add the decoder library to pom.xml using test scope, it can't be included always because its license is incompatible with our license.

> Also, the cache4.db file is the one I should be parsing, right? (Android version)

Yes.

streeg commented:
> > What should I do?
> 
> I will add the decoder library to pom.xml using test scope, it can't be included always because its license is incompatible with our license.
> 
> > Also, the cache4.db file is the one I should be parsing, right? (Android version)
> 
> Yes.

Thanks Nassif!

streeg commented:
Hey @lfcnassif, 
I have 'finished' Telegram (android) tests. For now, I used my own cache4.db, and I know that is not ideal. I'll try later to create a test account to populate the cache with info that is not personal. Also, later, when I create this account, I would appreciate if you could use it on some macOS platform and get me the cache files for IOS testing. I would do that myself, but I don't have the resources needed (nor an iPhone nor a macOS). 

Now for WhatsApp:
I'm starting WhatsApp parser tests, same thing as telegram, with my own files for testing and I know it's not ideal.
Once I have the test structure, it will be easier to adapt any other database file
I'm having a problem similar with what we had before: The parser is not finding the tables in the database. I have already tried what we did before and got no results. This is an example of what is happening: 

```
org.apache.tika.exception.TikaException: WAExtractorException Exception
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parseWhatsappMessages(WhatsAppParser.java:289)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parse(WhatsAppParser.java:190)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParserTest.testWhatsAppParser(WhatsAppParserTest.java:436)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: dpf.mg.udi.gpinf.whatsappextractor.WAExtractorException: org.sqlite.SQLiteException: [SQLITE_ERROR] SQL error or missing database (no such table: chat_list)
	at dpf.mg.udi.gpinf.whatsappextractor.ExtractorAndroid.extractChatList(ExtractorAndroid.java:93)
	at dpf.mg.udi.gpinf.whatsappextractor.Extractor.getChatList(Extractor.java:26)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parseWhatsappMessages(WhatsAppParser.java:284)
	... 21 more
Caused by: org.sqlite.SQLiteException: [SQLITE_ERROR] SQL error or missing database (no such table: chat_list)
	at org.sqlite.core.DB.newSQLException(DB.java:1012)
	at org.sqlite.core.DB.newSQLException(DB.java:1024)
	at org.sqlite.core.DB.throwex(DB.java:989)
	at org.sqlite.core.NativeDB.prepare_utf8(Native Method)
	at org.sqlite.core.NativeDB.prepare(NativeDB.java:134)
	at org.sqlite.core.DB.prepare(DB.java:257)
	at org.sqlite.jdbc3.JDBC3Statement.executeQuery(JDBC3Statement.java:66)
	at dpf.mg.udi.gpinf.whatsappextractor.ExtractorAndroid.extractChatList(ExtractorAndroid.java:70)
	... 23 more

```

I Also checked, and the provided files contains the tables. 
I don't know if is the way I am instantiating IItemSearcher and ItemInfo either. 
One last thing, how do I enable SHA256? I keep getting this error: 
`
17:25:59.582 [main] ERROR dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser - SHA-256 is disabled. WhatsAppParser needs it to link attachments to chats!
`

lfcnassif commented:
> I have 'finished' Telegram (android) tests. For now, I used my own cache4.db, and I know that is not ideal. I'll try later to create a test account to populate the cache with info that is not personal. Also, later, when I create this account, I would appreciate if you could use it on some macOS platform and get me the cache files for IOS testing. I would do that myself, but I don't have the resources needed (nor an iPhone nor a macOS).

I will ask the cellphone examiners if they have an iphone for testing.

> 
> Now for WhatsApp:
> I'm starting WhatsApp parser tests, same thing as telegram, with my own files for testing and I know it's not ideal.
> Once I have the test structure, it will be easier to adapt any other database file
> I'm having a problem similar with what we had before: The parser is not finding the tables in the database. I have already tried what we did before and got no results. This is an example of what is happening:

I will take a look.

lfcnassif commented:
@streeg I fixed the "no such table: chat_list" issue. Now a NPE is being thrown, but that is related to your IItemBase implementation. I suggest putting those empty IItemSearcher and IItemBase empty implementations somewhere else to be reused between tests.

About sha-256, it is used to find medias sent/received by chats. Seems you are not testing this for now, so don't worry. To test this properly, you will have to provide proper implementations of IItemSearcher/IItemBase and test sent/received files which hashes are referenced by the WA test database.

streeg commented:
I'm trying to use the option "setMergeDB" in the tests and whenever the parser tries to cast IItemBase to IItem throws this exception: 
```
java.lang.ClassCastException: dpf.mg.udi.gpinf.whatsappextractor.AbstractPkgTest$2 cannot be cast to iped3.IItem
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.checkIfIsMainDb(WhatsAppParser.java:315)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parse(WhatsAppParser.java:190)
	at dpf.mg.udi.gpinf.whatsappextractor.mergeDBTest.testWhatsAppParserMergeDBAndroid(mergeDBTest.java:53)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
``` 

This cast should happen automatically, right?

lfcnassif commented:
> This cast should happen automatically, right?

I think not, IItem extends IItemBase

> by the way, we are already covering 62% of the parsers!

Great! Thank you @streeg for all your work!

This PR is getting a bit large. I think you could target all remaining TODOs of existing tests (like date assertions, not sure if already fixed, etc) to merge this. And continue with new tests in a new PR, what do you think?

After finishing tests, I think you could focus your efforts on #43, what do you think?

streeg commented:
> Great! Thank you @streeg for all your work!

Thanks! It's been fun, and I've learned a lot about how each parser works individually.


> This PR is getting a bit large. I think you could target all remaining TODOs of existing tests (like date assertions, not sure if already fixed, etc) to merge this. And continue with new tests in a new PR, what do you think?

Looks good. I was actually thinking about it yesterday. This PR is really quite big, and maybe it would have been better to have separated it into little PRs from the start.
I still need to fix some little things in different tests, I think they can be done in new PRs.

> After finishing tests, I think you could focus your efforts on #43.

Sounds fun and challenging. I believe I will need some initial help or someone to ask questions during the creation of the new parsers. Once I finish most of the testing I will start investing in #43.

lfcnassif commented:
> I still need to fix some little things in different tests, I think they can be done in new PRs. I'll finish some things up until Friday, and then I'll close the PR. Then any smaller change I'll do in a new PR.

I think existing tests known issues should be addressed before merging. Do not close this, otherwise I won't be able to merge it, it will be closed automatically when merged.

> I believe I will need some initial help or someone to ask questions during the creation of the new parsers.

For sure, I'll try to help you of course.

lfcnassif reviewed (commented):
I think this maybe isn't the best approach to fix the issue. Tests and the application should work in any timezone.

streeg commented:
75c8041 when fixed will affect LNKShortcutParserTest and CertificateParserTest.

streeg commented:
> I think this maybe isn't the best approach to fix the date issues. Tests and the application should work in any timezone.

Oh, I don't know, but I have the feeling this fix is related with the testfile itself. I mean, the testfile was created in gmt-3 and the maven always tries to read it as gmt standard timezone. Right?

lfcnassif reviewed (changes requested):
please revert this and apply indentation fixes (if needed) separately, I can't see the logical changes mixed with space/tab changes

lfcnassif commented:
I also recommend stopping commiting many **test** code changes before knowing they will work.

lfcnassif commented:
And a "possible" solution to date assertion issues would be to test for Date objects equality, after converting strings to dates with a SimpleDateFormat with timezone properly set, maybe...

lfcnassif commented:
> I mean, the test file was created in gmt-3

I think this is not the issue

streeg commented:
> I also recommend stopping commiting many **test** code changes before knowing they will work. Remember this all will be merged to master in the future...

Apparently there was a GitHub problem during commits af888c1 ~ 94ddb0e. I thought it was something I changed, but it was actually GitHub itself. Hence, the large number of commits. I'm sorry. As for f8da144, 9fb7084 and 3fb6427 If not tested separately, during execution, when any of the tests breaks the workflow ends making it difficult to debug, so I separated into different commits.


> And a "possible" solution to date assertion issues would be to test for Date objects equality, after converting strings to dates with a SimpleDateFormat with timezone properly set (equal to the timezone used by the parser), maybe...

I'll take a look in that approach.





> please revert this commit [c1966e6](https://github.com/sepinf-inc/IPED/commit/c1966e63f4ef041e3bfcdef17e417033f5a07787) and apply indentation fixes (if needed) separately, I can't see the logical changes mixed with space/tab changes

I just added `-Duser.country=BR -Duser.language=pt` at line 185. Sorry for the mess.

lfcnassif commented:
> Apparently there was a GitHub problem

Sometimes CI fails to download big artifacts and the build fails, just try to run it again if that is shown in the build error log:
![image](https://user-images.githubusercontent.com/7276994/122775776-dec48500-d280-11eb-849d-d661ead47524.png)

> I just added `-Duser.country=BR -Duser.language=pt` at line 185

ok

streeg commented:
> Sometimes CI fails to download big artifacts and the build fails, just try to run it again if that is shown in the build error log:
> ![image](https://user-images.githubusercontent.com/7276994/122775776-dec48500-d280-11eb-849d-d661ead47524.png)

The error was  `GitHub Actions has encountered an internal error when running your job.` Clicking to re-run task option returned the same error. Initially I thought it was something I had changed in the pom. I just read later on GitHub forum that is literally an internal problem and happened with multiple users at the same time. I believe it has something to do with their servers. It is fixed now.

lfcnassif commented:
@streeg good afternoon. I saw some tests are hard coding some metadata key strings (like "userName" in VCardParserTest), I think it would be better to use the public static String key with the string value itself (ExtraProperties.xxx, Metadata.yyy, TikaCoreProperties.zzz...). Could you make a review in existing tests? If the String metadata key is already declared in parsers but as "private", you can simply change them to "public static final String"

streeg commented:
> @streeg good afternoon. I saw some tests are hard coding some metadata key strings (like "userName" in VCardParserTest), I think it would be better to use the public static String key with the string value itself (ExtraProperties.xxx, Metadata.yyy, TikaCoreProperties.zzz...). Could you make a review in existing tests? If the String metadata key is already declared in parsers but as "private", you can simply change them to "public static final String"

I will do that.

streeg commented:
@lfcnassif I just remembered why the metadata info from some tests are 'hard coded' assertions. I tried to fill the metadata info in my Abstract Package, but for some reason it was not receiving any information. Since some metadata fields are filled multiple times, I use the abstract package to create a tracker that receives all metadata info as a list of properties. That way, I can access them by index. For VCardParser and some others parsers, these fields are not filled, and I don't know why. 

![2021-06-24 15_46_04-eclipse-workspace - iped-parsers-impl_src_test_java_dpf_mg_udi_gpinf_vcardparser](https://user-images.githubusercontent.com/22523306/123316656-7933fb00-d503-11eb-9d37-923723253771.png)
![2021-06-24 15_46_35-](https://user-images.githubusercontent.com/22523306/123316664-7b965500-d503-11eb-8250-42f24a3b5e54.png)

streeg commented:
Also, there are some metadata parsers fields there are not from ExtraProperties, TikaCore or Metadata... In these cases, it is okay to leave hard coded?

lfcnassif commented:
> @lfcnassif I just remembered why the metadata info from some tests are 'hard coded' assertions. I tried to fill the metadata info in my Abstract Package, but for some reason it was not receiving any information. Since some metadata fields are filled multiple times, I use the abstract package to create a tracker that receives all metadata info as a list of properties. That way, I can access them by index. For VCardParser and some others parsers, these fields are not filled, and I don't know why.

I will try to take a look when possible.



> Also, there are some metadata parsers fields there are not from ExtraProperties, TikaCore or Metadata... In these cases, it is okay to leave hard coded?

There is also BasicProperties. Take a look at parsers code to see if they use some static String key declared somewhere else. If the strings are hardcode into the parser, leave it as is now, these parsers should be improved. If they use some declared string, the test should also use, I think.

lfcnassif commented:
Maybe you tried to commit to the protected master branch instead of ParserTests branch?

streeg commented:
> Maybe you tried to commit to the protected master branch instead of ParserTests branch?

Already checked that. Was trying to commit to ParserTests. I still have the commit permissions, right?

lfcnassif commented:
> I still have the commit permissions, right?

I didn't change anything, but checked and seems ok. I have just pushed to a feature branch without issues...

streeg commented:
It seems it solved itself. Probably some GitHub bug. Thanks for the attention!

lfcnassif commented:
Hi @streeg, thanks for all your work on this. I ask you to do a final review on existing tests, so I will review them later before merging, this is becoming too large...

Please I ask you to include new tests in another PR, thank you!

streeg commented:
> Hi @streeg, thanks for all your work on this. I ask you to do a final review on existing tests, so I will review them later before merging, this is becoming too large...
> 
> Please I ask you to include new tests in another PR, thank you!

I'll do a final review!

lfcnassif commented:
MESSAGE_IS_ATTACHMENT is not meant to flag emails with attachments, but the attachments extracted from emails.

streeg commented:
> @streeg the last commit is wrong. MESSAGE_IS_ATTACHMENT should be set in attachments metadata only, not in the MSG metadata itself. Please revert your changes on MSGParser and adjust the new test accordingly, thank you!


When setted where it was being set, even the messages with attachments were not changing the MESSAGE_IS_ATTACHMENT property to TRUE. 
But I'll change it back. 
This still odd to me.

lfcnassif commented:
> But I'll change it back.

thanks @streeg, I implemented a proper test on ccf479f, take a look.

lfcnassif reviewed (commented):
Thanks @streeg

lfcnassif commented:
@streeg also please finish some tests without any assertion, like LibpffPSTParserTest, thank you.

streeg commented:
I just noticed that if you try to build the project on Windows System where the IPED folder path contains spaces some tests break with exception "can't find path to this file". 

`java.io.FileNotFoundException: C:\Users\Guilherme%20Andreuce\git\IPED\iped-parsers\iped-parsers-impl\target\test-classes\test-files\test_security (O sistema não pode encontrar o caminho especificado)` 

I read it [here](https://stackoverflow.com/questions/10536183/resource-files-not-found-from-junit-test-cases#comment13631623_10536183) that could be the pathing. I moved the IPED folder from 'users/git' to 'C:/git' and seems to have solved the issue.

lfcnassif commented:
So please test the .toURI() recommendation in the link and apply to all tests if it works, thanks!

streeg commented:
> So please test the .toURI() recommendation in the link and apply to all tests if it works, thanks!

I'll test after closing the streams

lfcnassif commented on iped-parsers/iped-parsers-impl/src/test/java/dpf/mg/udi/gpinf/vcardparser/VCardParserTest.java:
```diff
@@ -197,12 +196,13 @@ public void testVCardCompleteHTMLToString() throws IOException, SAXException, Ti
-        ParseContext context = new ParseContext();
-        parser.getSupportedTypes(context);
         try(InputStream stream = getStream("test-files/test_contactsCompleteWithoutThumbHTMLToString.vcf")){
+	        File file = new File(
```
@streeg this method has many errors: stream closed twice, stream.toString() and seems it is trying to overwrite the testing file.

streeg replied:
Just learned that try block closes InputStream automatically. Didn't knew about the java.lang.AutoCloseable feature. Sorry, I already removed them. About VCardParserTest, my mistake with the stream, I might have mispelled by accident.

lfcnassif replied:
Thank you!

lfcnassif reviewed (changes requested):
Actually, all methods in this class are closing the stream twice. Please review all other tests and fix, thanks.

lfcnassif commented:
@streeg I merged this. Please finish all TODOs here and start other tests in another branch and PR, thank you.
